state/traverse.go

```diff
@@ -7,27 +7,59 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
+// functor that will be get called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) (bool, error)
+
 // Traverse traverses a chain segment beginning with the start block (inclusive)
-// and ending with the end block (inclusive). Blocks are traversed in reverse
+// Blocks are traversed in reverse
 // height order, meaning the end block must be an ancestor of the start block.
 // The callback is called for each block in this segment.
-func Traverse(headers storage.Headers, start, end flow.Identifier, callback func(header *flow.Header) error) error {
-
-	nextID := start
+// Return value of callback is used to decide if it should continue or not.
+func Traverse(headers storage.Headers, startBlockID flow.Identifier, visitor onVisitBlock) error {
+	ancestorID := startBlockID
```

Comment:
I find the naming of `ancestor` somewhat confusing, because we are processing the block itself, not its parent. Would suggest to rename:
* `ancestor` -> `block`
* `ancestorID` -> `blockID`